File: Grind75/31-LevelOrderTraversal/cpp/LevelOrderTraversal.cpp
#include "LevelOrderTraversal.h"
#include <queue>
#include <utility>
#include <vector>

using namespace std;
using namespace BST;

vector<vector<int>> LevelOrderTraversal::levelOrder(TreeNode* root) {
  queue<pair<TreeNode*, int>> nodeStack;
  nodeStack.push(make_pair(root, 0));
  vector<vector<int>> result;

  while (!nodeStack.empty()) {
    pair<TreeNode*, int> &front = nodeStack.front();
    TreeNode *node = front.first;
    int level = front.second;
    nodeStack.pop();
    if (!node)
      continue;
    if (result.size() <= static_cast<size_t>(level))
      result.push_back({node->val});
    else
      result[level].push_back(node->val);
    nodeStack.push(make_pair(node->left, level+1));
    nodeStack.push(make_pair(node->right, level+1));
  }
  return result;
}

vector<vector<int>> LevelOrderTraversal::levelOrderRecursive(TreeNode* root) {
  vector<vector<int>> result;
  helper(root, 0, result);
  return result;
}

void LevelOrderTraversal::helper(TreeNode* node, int level, vector<vector<int>> &result) {
  if (!node)
    return;
  if (static_cast<size_t>(level) >= result.size())
    result.push_back({node->val});
  else
    result[level].push_back(node->val);
  helper(node->left, level+1, result);
  helper(node->right, level+1, result);
}

File: Grind75/31-LevelOrderTraversal/python/LevelOrderTraversal.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')

from typing import Optional, List
from Tree import *
from collections import deque
from PrettyPrint import *

class Solution(object):
  def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
    if (not root):
      return []
    queue = deque()
    result: List[List[int]] = []
    queue.appendleft((root, 0))

    while len(queue) != 0:
      node: TreeNode
      level: int
      (node, level) = queue.pop()
      if (not node):
        continue
      if len(result) <= level:
        result.append([node.val])
      else:
        result[level].append(node.val)
      queue.appendleft((node.left, level+1))
      queue.appendleft((node.right, level+1))

    return result

if __name__ == '__main__':
  soln = Solution()
  test = makeBST([4, 2, 1, 3, 6, 7])

  printTree(test)
  prettyPrint2D(soln.levelOrder(test))

